File: context/data/apache_arrow/sample/client/test_client.py
import requests
import pyarrow as pa
import pyarrow.ipc as ipc
import io
import datetime

BASE_URL = "http://localhost:8080"

def test_upload():
    print("Testing /upload...")
    # Create sample data
    data = [
        pa.array([10, 11, 12], type=pa.int32()),
        pa.array([100.1, 200.2, 300.3], type=pa.float64()),
        pa.array([datetime.datetime.now() for _ in range(3)], type=pa.timestamp('ns'))
    ]
    schema = pa.schema([
        ('id', pa.int32()),
        ('value', pa.float64()),
        ('timestamp', pa.timestamp('ns'))
    ])
    batch = pa.RecordBatch.from_arrays(data, schema=schema)
    
    sink = io.BytesIO()
    with ipc.new_stream(sink, schema) as writer:
        writer.write_batch(batch)
    
    resp = requests.post(f"{BASE_URL}/upload", data=sink.getvalue(), headers={'Content-Type': 'application/vnd.apache.arrow.stream'})
    print(f"Status: {resp.status_code}, Body: {resp.text}")

def read_arrow_stream(content):
    with ipc.open_stream(content) as reader:
        schema = reader.schema
        print(f"Schema: {schema}")
        for batch in reader:
            print(f"Batch has {batch.num_rows} rows and {batch.num_columns} columns")
            print(batch.to_pydict())

def test_download():
    print("\nTesting /download...")
    resp = requests.get(f"{BASE_URL}/download")
    print(f"Status: {resp.status_code}")
    if resp.status_code == 200:
        read_arrow_stream(resp.content)

def test_query():
    print("\nTesting /query...")
    query = "SELECT * FROM measurements WHERE value > 150"
    resp = requests.post(f"{BASE_URL}/query", data=query, headers={'Content-Type': 'text/plain'})
    print(f"Status: {resp.status_code}")
    if resp.status_code == 200:
        read_arrow_stream(resp.content)

def test_stats():
    print("\nTesting /stats...")
    resp = requests.get(f"{BASE_URL}/stats")
    print(f"Status: {resp.status_code}")
    if resp.status_code == 200:
        read_arrow_stream(resp.content)

def test_echo():
    print("\nTesting /echo...")
    data = [
        pa.array([1, 2], type=pa.int32()),
        pa.array([1.1, 2.2], type=pa.float64()),
        pa.array([datetime.datetime.now() for _ in range(2)], type=pa.timestamp('ns'))
    ]
    schema = pa.schema([
        ('id', pa.int32()),
        ('value', pa.float64()),
        ('timestamp', pa.timestamp('ns'))
    ])
    batch = pa.RecordBatch.from_arrays(data, schema=schema)
    
    sink = io.BytesIO()
    with ipc.new_stream(sink, schema) as writer:
        writer.write_batch(batch)
    
    resp = requests.post(f"{BASE_URL}/echo", data=sink.getvalue(), headers={'Content-Type': 'application/vnd.apache.arrow.stream'})
    print(f"Status: {resp.status_code}")
    if resp.status_code == 200:
        read_arrow_stream(resp.content)

if __name__ == "__main__":
    try:
        test_upload()
        test_download()
        test_query()
        test_stats()
        test_echo()
    except requests.exceptions.ConnectionError:
        print("Error: Could not connect to the server. Make sure main.go is running on http://localhost:8080")

File: context/data/apache_arrow/sample/client/test_client.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/apache/arrow/go/v14/arrow"
	"github.com/apache/arrow/go/v14/arrow/array"
	"github.com/apache/arrow/go/v14/arrow/ipc"
	"github.com/apache/arrow/go/v14/arrow/memory"
)

var pool = memory.NewGoAllocator()

func main() {
	time.Sleep(2 * time.Second) // Wait for server to start if running concurrently
	
	// Create sample data
	schema := arrow.NewSchema(
		[]arrow.Field{
			{Name: "id", Type: arrow.PrimitiveTypes.Int32},
			{Name: "value", Type: arrow.PrimitiveTypes.Float64},
			{Name: "timestamp", Type: arrow.FixedWidthTypes.Timestamp_ns},
		},
		nil,
	)

	b := array.NewRecordBuilder(pool, schema)
	defer b.Release()

	b.Field(0).(*array.Int32Builder).AppendValues([]int32{1, 2, 3}, nil)
	b.Field(1).(*array.Float64Builder).AppendValues([]float64{10.5, 20.7, 30.2}, nil)
	b.Field(2).(*array.TimestampBuilder).AppendValues([]arrow.Timestamp{
		arrow.Timestamp(time.Now().UnixNano()),
		arrow.Timestamp(time.Now().UnixNano()),
		arrow.Timestamp(time.Now().UnixNano()),
	}, nil)

	rec := b.NewRecord()
	defer rec.Release()

	var buf bytes.Buffer
	writer := ipc.NewWriter(&buf, ipc.WithSchema(schema))
	if err := writer.Write(rec); err != nil {
		log.Fatal(err)
	}
	writer.Close()

	// 1. Test /upload
	fmt.Println("Testing /upload...")
	resp, err := http.Post("http://localhost:8080/upload", "application/vnd.apache.arrow.stream", &buf)
	if err != nil {
		log.Fatal(err)
	}
	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("Status: %s, Body: %s\n", resp.Status, body)

	// 2. Test /download
	fmt.Println("\nTesting /download...")
	resp, err = http.Get("http://localhost:8080/download")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Status: %s\n", resp.Status)
	readArrow(resp.Body)

	// 3. Test /query
	fmt.Println("\nTesting /query...")
	resp, err = http.Post("http://localhost:8080/query", "text/plain", bytes.NewBufferString("SELECT * FROM measurements WHERE value > 20"))
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Status: %s\n", resp.Status)
	readArrow(resp.Body)

	// 4. Test /stats
	fmt.Println("\nTesting /stats...")
	resp, err = http.Get("http://localhost:8080/stats")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Status: %s\n", resp.Status)
	readArrow(resp.Body)

	// 5. Test /echo
	fmt.Println("\nTesting /echo...")
	buf.Reset()
	writer = ipc.NewWriter(&buf, ipc.WithSchema(schema))
	writer.Write(rec)
	writer.Close()
	resp, err = http.Post("http://localhost:8080/echo", "application/vnd.apache.arrow.stream", &buf)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Status: %s\n", resp.Status)
	readArrow(resp.Body)
}

func readArrow(r io.Reader) {
	reader, err := ipc.NewReader(r, ipc.WithAllocator(pool))
	if err != nil {
		fmt.Printf("Error creating reader: %v\n", err)
		return
	}
	defer reader.Release()

	for reader.Next() {
		rec := reader.Record()
		fmt.Printf("Record has %d rows and %d columns\n", rec.NumRows(), rec.NumCols())
		fmt.Printf("Schema: %v\n", rec.Schema())
		// Print first row as example
		for i := 0; i < int(rec.NumCols()); i++ {
			fmt.Printf("Col %d (%s): %v\n", i, rec.Schema().Field(i).Name, rec.Column(i))
		}
	}
}
